### PR TITLE
Support encoder / decoder class

### PR DESCRIPTION
**Pull request by lvdgraaff**
Add `cls` argument to `open()`, `Writer()` and `Reader()`, similarly as in `json.dump`and  `json.load`. 
This can, for example, be used in combination with the NumpyEncoder class from the [numpyencoder](https://pypi.org/project/numpyencoder/) module.

Example code:
```py
import jsonlines
import numpy as np
from numpyencoder import NumpyEncoder

with jsonlines.open('example.jsonl', 'w', cls=NumpyEncoder) as l: 
    l.write(dict(test=np.array([1,2,3]))) 
```

This change is intended as a usability improvement over using 
```py
...
with jsonlines.open('test.jsonl', mode='a', dumps=NumpyEncoder().encode) as l
...
```

## Implementation notes
* The local variable `cls` in `open` had to be renamed to `instance_cls` to prevent confusion. 
* No additional arguments can be supplied to the encoder/decoder, where json.dump, jump.load passes all additional kwargs to the constructor of the encoder/decoder.

### CODE REVIEW

**Comment by wbolster** [comment]
🫤 i'm a bit confused about this. does this solve any problem that cannot be solved using the current `loads=` and `dumps=` keyword args?

as your own example shows, it's just a few extra characters, but allows passing arbitrary arguments to the encoder/decoder instance, and the resulting code makes it very clear what happens.

a `cls=` arg that sometimes applies to writing (encoding) and sometimes applies to reading (decoding) seems confusing to me as well. and the type annotations aren't nice either: `type` is very generic, compared to the rather strict signatures for `loads=` and `dumps=` argument values:

``` python
LoadsCallable = Callable[[Union[str, bytes]], Any]
DumpsCallable = Callable[[Any], Union[str, bytes]]
```

in python, functions (and other callables like methods on a class instance) are first-class objects, so going the indirect route via a class that has to be instantiated (somehow), only to call a single method on them, the name and signature of which depends on whether this is a reader or a writer... seems to add implicitness and complexity for no gain? but maybe i'm missing something? 🤔

**Comment by lvdgraaff** [comment]
Hi Wouter, 

Thanks for your fast and thorough response. Let met answer you point by point.

> i'm a bit confused about this. does this solve any problem that cannot be solved using the current loads= and dumps= keyword args?

This change is not an added functionality but rather an usability improvements This change is not so much about more convenience - indeed, it just a few characters shorter - but to provide familiarity. To clarify this: the json module provides the `cls` argument for having a custom encoder/decoder, see e.g., [here](https://stackoverflow.com/questions/26646362/numpy-array-is-not-json-serializable) or [here](https://pypi.org/project/numpyencoder/).

```py
import numpy as np
import json
from numpyencoder import NumpyEncoder

with open(‘example.json’, ‘w’) as f:
    json.dump(f, dict(test=np.array([1,2,3]), cls=NumpyEncoder)
```

> a cls= arg that sometimes applies to writing (encoding) and sometimes applies to reading (decoding) seems confusing to me as well. and the type annotations aren't nice either: type is very generic, compared to the rather strict signatures for loads= and dumps= argument values:

This is a legit remark, and has been addressed in 93a29ffb35e41c352368265fbee651d7b1c89267 where the Writer and Reader now have a strict type hint and the use of the cls argument in the `open` method is clarified in the over `@overload` section.

> in python, functions (and other callables like methods on a class instance) are first-class objects, so going the indirect route via a class that has to be instantiated (somehow), only to call a single method on them, the name and signature of which depends on whether this is a reader or a writer... seems to add implicitness and complexity for no gain? 

I fully agree with you here. However, this is a design choice that has been made by the creators of the json module and that users may be familiar to. Also, there might be various custom encoders around that are intended and documented to be used in this way.

Anyhow, thank you very much for developing and maintaining jsonlines, it's highly appreciated.

Best,
Leon

**Comment by wbolster** [comment]
thanks for the reply. i especially appreciate things like this: 🙏


> Anyhow, thank you very much for developing and maintaining jsonlines, it's highly appreciated.

that said, the more i think about this, the more you convince me this PR is actually not an improvement at all – sorry 🙃

unfortunately the type annotations like the ones you've added, such as

``` python
    _cls : Type[json.JSONDecoder] = attr.ib(default=json.JSONDecoder, kw_only=True)
```

… may seem tighter and better at first sight, but are actually wrong on a more abstract level.

the interfaces needed by the `jsonlines` library are either `loads(s)` and `dumps(obj)`, _not_ some decoder/encoder class with constructor interfaces and some methods. in fact, the mere existence of a decoder/encoder class for json decoding/encoding is an implementation detail that is deliberately not exposed by the `jsonlines` API, and that is deliberate.

the whole point of `dumps=` and `loads=` is to allow plugging in different json encoders/decoders, e.g. ultrajson/`ujson` (https://github.com/ultrajson/ultrajson). those different encoders/decoders may not even _have_ classes to do encoding/decoding, e.g. `ujson` does not have one, not even under the hood.

for decoding, `jsonlines` uses no (explicit) `json.JSONDecoder` at all, though `json.loads()` uses one under the hood – an implementation detail of the `json` module that `jsonlines` does not care about.

for encoding, the only usage of `json.JSONEncoder` is strictly internal and only used as an optimization: creating a `JSONEncoder` only _once_ for the lifetime of a `jsonlines.Writer` is faster than calling `json.dumps()` repeatedly, and even more so if non-standard flags are used. `jsonlines` always used custom flags: `ensure_ascii=False` and potentially more, depending on the `compact=` flag. the result is that case `jsonlines` will never benefit from the (internal and invisible) cached `_default_encoder` used by the stdlib `json` module.  for reference, here is the implementation `json.dumps()` from the python stdlib:

``` python
def dumps(obj, *, skipkeys=False, ensure_ascii=True, check_circular=True,
        allow_nan=True, cls=None, indent=None, separators=None,
        default=None, sort_keys=False, **kw):
    # cached encoder
    if (not skipkeys and ensure_ascii and
        check_circular and allow_nan and
        cls is None and indent is None and separators is None and
        default is None and not sort_keys and not kw):
        return _default_encoder.encode(obj)
    if cls is None:
        cls = JSONEncoder
    return cls(
        skipkeys=skipkeys, ensure_ascii=ensure_ascii,
        check_circular=check_circular, allow_nan=allow_nan, indent=indent,
        separators=separators, default=default, sort_keys=sort_keys,
        **kw).encode(obj)
```

long story short: i'm gonna close this as `wontfix`

**Comment by lvdgraaff** [comment]
Thanks for this good conversation and the convincing argumentation.

How would you feel about adding an example to the doc on using a custom encoder, with the Numpy encoder as example? I could do a proposal if you're interested.

**Comment by wbolster** [comment]
sure, doc improvements welcome
